tools/storage.py

```diff
@@ -41,29 +44,120 @@ def __init__(self, batch):
+            result = int(result)
+            result = (result >> 1) ^ -(result & 1)
+            return (result, pos)
+        shift += 7
+        if shift >= 64:
+            raise _DecodeError('Too many bytes when decoding varint.')
+
+class Record:
+    def __init__(self, size, record_attr, timestamp_delta, offset_delta, key, headers, value):
+        self.size = size
+        self.record_attr = record_attr
+        self.timestamp_delta = timestamp_delta
+        self.offset_delta = offset_delta
+        self.key = key
+        self.headers = headers
+        self.value = value 
+
+    def __str__(self):
+        return "{}:{}".format(self.key.decode("utf-8"), self.value.decode("utf-8"))
+
+
+    @staticmethod
+    def record_from_bytes(b):
+        (size, v) = read_varlong(b, 0)
+        
+        record_attr = b[v]
+        v = v + 1
+
+        (timestamp_delta, v) = read_varlong(b, v)
+        (offset_delta, v) = read_varlong(b, v)
+
+        (key_length, v) = read_varlong(b, v)
+        key = b[v:v+key_length]
+        v = v + key_length
+
+        (value_length, v) = read_varlong(b, v)
+        value = b[v:v+value_length]
+        v = v + value_length
+
+        (header_count, v) = read_varlong(b, v)
+        headers = []
+        for i in range(0, header_count):
+            (header_length, v) = read_varlong(b, v)
+            header = b[v:v+header_length]
+            v = v + header_length
+            headers.append(header)
+
+        return Record(size, record_attr, timestamp_delta, offset_delta, key, headers, value)
+
+    @staticmethod
+    def from_bytes(b, count):
+        records = []
+        for i in range(0, count):
+            record = Record.record_from_bytes(b)
+            b = b[:record.size]
+            records.append(record)
+        return records
+
+
 class Batch:
     def __init__(self, index, header, records):
         self.index = index
         self.header = header
+        self.records = records
 
+    def verify_crc(self):
         header_crc_bytes = struct.pack(
             "<" + HDR_FMT_RP_PREFIX_NO_CRC + HDR_FMT_CRC, *self.header[1:])
-        header_crc = crc32c.crc32(header_crc_bytes)
+        header_crc = crc32c.crc32c(header_crc_bytes)
         if self.header.header_crc != header_crc:
             raise CorruptBatchError(self)
 
-        crc = crc32c.crc32(self._crc_header_be_bytes())
-        crc = crc32c.crc32(records, crc)
+        crc = crc32c.crc32c(self._crc_header_be_bytes())
+        crc = crc32c.crc32c(self.records, crc)
         if self.header.crc != crc:
             raise CorruptBatchError(self)
 
     def last_offset(self):
-        return self.header.base_offset + self.header.record_count - 1
+        return self.header.base_offset + self.header.delta
 
     def _crc_header_be_bytes(self):
         # encode header back to big-endian for crc calculation
         return struct.pack(">" + HDR_FMT_CRC, *self.header[5:])
 
+    def parse_records(self):
+        batch = self.records
```

Comment:
don't we need to test if zstd is in the header? we can simply log as an alternative or raise an exception if 

```
if header.compression != zstd: 
   raise UnsupportedCompression("We currently only support zstd; please submit a patch :D")

```

Reply:
I found what looks like a mapping for compression types to batch header attrs
https://github.com/vectorizedio/redpanda/blob/dev/src/v/model/compression.h#L35

I went ahead and implemented lz4 and gzip since they seemed straight forward, but they aren't tested. I moved the imports into the conditional so if you aren't using --dump you won't need the libraries installed.